src/main/java/com/google/devtools/build/lib/profiler/CommandProfilerModule.java

```diff
@@ -0,0 +1,121 @@
+// Copyright 2023 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.profiler;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.flogger.GoogleLogger;
+import com.google.devtools.build.lib.events.Event;
+import com.google.devtools.build.lib.events.Reporter;
+import com.google.devtools.build.lib.runtime.BlazeModule;
+import com.google.devtools.build.lib.runtime.Command;
+import com.google.devtools.build.lib.runtime.CommandEnvironment;
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.common.options.Option;
+import com.google.devtools.common.options.OptionDocumentationCategory;
+import com.google.devtools.common.options.OptionEffectTag;
+import com.google.devtools.common.options.OptionsBase;
+import java.time.Duration;
+import javax.annotation.Nullable;
+import one.profiler.AsyncProfiler;
+
+/**
+ * Bazel module to record pprof-compatible profiles for single invocations.
+ */
+public class CommandProfilerModule extends BlazeModule {
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  private static final Duration PROFILING_INTERVAL = Duration.ofMillis(10);
+
+  @Nullable
+  private static final AsyncProfiler profiler;
+
+  static {
+    AsyncProfiler profilerInstance = null;
+    try {
+      profilerInstance = AsyncProfiler.getInstance();
+    } catch (Exception e) {
+      // Loading the JNI must be allowed to fail, as we might be running on an unsupported platform.
+      logger.atWarning().log("Failed to load async_profiler JNI: %s", e);
+    }
+    profiler = profilerInstance;
+  }
+
+  public static final class Options extends OptionsBase {
+
+    @Option(
+        name = "experimental_command_profile",
+        defaultValue = "false",
```

Comment:
Do we want to make this a boolean option or already some enum-based option where CPU is currently the only value? Then it will be easier to add other modes in the future.

Reply:
Yes, I think this will eventually be an enum, but I'd rather not do that until I've actually figured out how to get the other profile types out of async-profiler and convert them to pprof. I think it's fine to change the syntax later since it's marked experimental?

Reply:
Ok, sounds reasonable.